rct229/rule_engine/rct_report.py

```diff
@@ -0,0 +1,139 @@
+        self.date_run = str(datetime.utcnow())
+        self.schema_version = "0"
+        self.ruleset_report_file = "report.txt"
+        self.ruleset_outcome = {
+            RCTOutcomeLabel.PASS: 0,
+            RCTOutcomeLabel.FAILED: 0,
+            RCTOutcomeLabel.UNDETERMINED: 0,
+            RCTOutcomeLabel.NOT_APPLICABLE: 0,
+        }
+
+    def generate(self, rct_outcome, report_dir):
+        """
+        generate rule report. - do not modify/override this function
+        This method also orchestrates the high-level workflow for any rule report
+        1. initialize_ruleset_report(): initialize a ruleset data structure
+        2. generate_rule_report(): generate a report for one rule
+        3. calculate_rule_outcome(): calculates the rule outcome
+        4. add_rule_to_ruleset_report(): append the rule outcome to the ruleset report
+        5. save_ruleset_report(): save the ruleset report to the report_dir
+
+        Parameters
+        ----------
+        rct_outcome: dict dictionary outcome generated from RCT engine
+        report_dir: str report file that saves the report output.
+
+        Returns - Save the output to a folder
+        -------
+
+        """
+        invalid_msg = rct_outcome["invalid_rmrs"]
+        outcomes = rct_outcome["outcomes"]
+        if invalid_msg:
+            print(f"Invalid RMDs: {str(invalid_msg)}\n")
+            pass
+        else:
+            ruleset_report = self.initialize_ruleset_report()
+            for outcome in outcomes:
+                rule_outcome_dict = {
+                    RCTOutcomeLabel.PASS: 0,
+                    RCTOutcomeLabel.FAILED: 0,
+                    RCTOutcomeLabel.UNDETERMINED: 0,
+                    RCTOutcomeLabel.NOT_APPLICABLE: 0,
+                }
+                rule_report = self.generate_rule_report(outcome, rule_outcome_dict)
+                rule_outcome = self.calculate_rule_outcome(rule_outcome_dict)
+                self.add_rule_to_ruleset_report(
+                    ruleset_report, rule_report, rule_outcome
+                )
+            report_dir = os.path.join(report_dir, self.ruleset_report_file)
+            self.save_ruleset_report(ruleset_report, report_dir)
+
+    def generate_rule_report(self, rule_outcome, outcome_dict):
+        """
+        Function to generate a rule's report
+
+        Parameters
+        ----------
+        rule_outcome: json contains a rule's raw report
+        outcome_dict: outcome dictionary, default is
+
+        Returns: rule re
+        -------
+
+        """
+        return rule_outcome
+
+    def initialize_ruleset_report(self):
+        """
+        Initialize a data structure for generating a ruleset report
+        default is list
+
+        Returns: data structure, default is list
+        -------
+
+        """
+        return list()
+
+    def add_rule_to_ruleset_report(self, ruleset_report, rule_report, rule_outcome):
+        """
+        Add rule to the ruleset report
```

Comment:
I assume that subclasses MUST implement this method.  It would be good to add a comment to that effect.
Same for the `save_ruleset_report` method.